### PR TITLE
[CELEBORN-1452] Master follower node metadata is out of sync after installing snapshot

### PR DESCRIPTION
**Pull request by leixm**
### What changes were proposed in this pull request?
Fix Master follower node metadata is out of sync after installing snapshot

### Why are the changes needed?
Follower node metadata is out of sync, when a master-slave switchover occurs, there are major risks to the stability of the cluster.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
UT.

### CODE REVIEW

**Comment by leixm** [comment]
```
follower log
24/05/23 03:08:02,421 INFO [1@group-47BEDE733167-StateMachineUpdater] StateMachine: Reinitializing state machine.
24/05/23 03:08:02,421 INFO [1@group-47BEDE733167-StateMachineUpdater] StateMachine: obsolete snapshot provided: (t:28, i:1185003348)

24/05/23 03:06:53,439 WARN [master-state-machine-executor] AbstractMetaManager: Celeborn cluster estimated partition size changed from 129.2 MiB to 110.5 MiB
24/05/23 03:08:01,425 INFO [1@group-47BEDE733167-StateMachineUpdater] StateMachine: Taking a snapshot to file /mnt/ssd/0/celeborn-ratis/c5196f6d-2c34-3ed3-8b8a-47bede733167/tmp/raft_snapshot_1716404881416_5912151569428133017.dat.
24/05/23 03:08:01,428 INFO [1@group-47BEDE733167-StateMachineUpdater] StateMachine: Saving digest 184db806c56d5780f92c98fd7d3fb35a for snapshot file /mnt/ssd/0/celeborn-ratis/c5196f6d-2c34-3ed3-8b8a-47bede733167/sm/snapshot.28_1185003348.
24/05/23 03:08:01,428 INFO [1@group-47BEDE733167-StateMachineUpdater] StateMachine: Renaming a snapshot file /mnt/ssd/0/celeborn-ratis/c5196f6d-2c34-3ed3-8b8a-47bede733167/tmp/raft_snapshot_1716404881416_5912151569428133017.dat to /mnt/ssd/0/celeborn-ratis/c5196f6d-2c34-3ed3-8b8a-47bede733167/sm/snapshot.28_1185003348.
24/05/23 03:08:02,421 INFO [1@group-47BEDE733167-StateMachineUpdater] StateMachine: Reinitializing state machine.
24/05/23 03:08:02,421 INFO [1@group-47BEDE733167-StateMachineUpdater] StateMachine: obsolete snapshot provided: (t:28, i:1185003348)

leader log
24/05/23 03:08:02,117 INFO [2@group-47BEDE733167-StateMachineUpdater] StateMachine: Taking a snapshot to file /mnt/ssd/0/celeborn-ratis/c5196f6d-2c34-3ed3-8b8a-47bede733167/tmp/raft_snapshot_1716404882108_1183932657463440669.dat.
24/05/23 03:08:02,120 INFO [2@group-47BEDE733167-StateMachineUpdater] StateMachine: Saving digest 232b855f8ca673b7a2c38a8ec90c6306 for snapshot file /mnt/ssd/0/celeborn-ratis/c5196f6d-2c34-3ed3-8b8a-47bede733167/sm/snapshot.28_1185003493.
24/05/23 03:08:02,121 INFO [2@group-47BEDE733167-StateMachineUpdater] StateMachine: Renaming a snapshot file /mnt/ssd/0/celeborn-ratis/c5196f6d-2c34-3ed3-8b8a-47bede733167/tmp/raft_snapshot_1716404882108_1183932657463440669.dat to /mnt/ssd/0/celeborn-ratis/c5196f6d-2c34-3ed3-8b8a-47bede733167/sm/snapshot.28_1185003493.
```

After reinitializing, follower out of sync.
meta

There are some variables find from follower jvm heap

1. StateMachineUpdater.appliedIndex   = 1185003348
2. SegmentedRaftLog.committedIndex = 1188743857
3. SegmentedRaftLog startIndex = 1185003494, endIndex = 1188743857 
4. StateMachineUpdater.snapshotIndex = 1185003348
5. ServerState.latestedInstalledSnapshot=1185003493

 

1. When will reinitializing be triggered?
It is triggered after leader sends a snapshot to the follower and finish installing.

 

2. When does the leader send a snapshot to the followers in the running state?
When the follower's nextIndex < the leader's startIndex, that is, the log required by the follower no longer exists in the leader.
    2.1 Why does the log not exist in the leader?
    There is a certain synchronization delay between the leader and the folower. When the leader takes the snapshot first, it will purge its own segment log after taking it. At this time, the log with index <= snapshot index can be purged, because there is already a snapshot. At this time, the above judgment will be Take effect and then send the snapshot

 

org.apache.ratis.server.leader.LogAppender#shouldInstallSnapshot

```
final long followerNextIndex = getFollower().getNextIndex();
if (followerNextIndex < getRaftLog().getNextIndex()) {
final long logStartIndex = getRaftLog().getStartIndex();
if (followerNextIndex < logStartIndex || (logStartIndex == RaftLog.INVALID_LOG_INDEX && snapshot != null))

{ return snapshot; }
}
```

3. Why is the snapshot sent by the leader to the follower not load correctly?
When the follower receives the snapshot, it will write to snapshotDir, but it will not refresh the latestSnapshot, so it cannot see this snapshot when reinitialize. snapshot.getTermIndex().compareTo(getLastAppliedTermIndex()) <= 0 must be true

 

4. Why is the final metadata of the follower out of sync?
After the leader sends the snapshot and receives the success status, it will update the follower's nextIndex to the sent snapshot last index + 1. The leader thinks that the follower has correctly installed snapsohot, but in fact it has not, which ultimately causes the follower to miss a section of the log. If the follower cannot find the log, it will lose synchronization forever.

 
org.apache.ratis.server.impl.StateMachineUpdater#applyLog
```
final LogEntryProto next = raftLog.get(nextIndex);
if (next != null) {xxx}
else {
LOG.debug("{}: logEntry {} is null. There may be snapshot to load. state:{}",
this, nextIndex, state);
break;
}
```

 

5. How do we solve this problem?
We need to modify the SimpleStateMachineStorage currently used by celeborn. When the snapshot install is completed, we should select the latest snapshot from the local disk directory and update the latestSnapshot variable.

**Comment by leixm** [comment]
@AngersZhuuuu  @waitinfuture  @SzyWilliam  @pan3793  Can you help review?

**Comment by leixm** [comment]
UT shows how to reproduce

**Comment by AngersZhuuuu** [comment]
From ArithmeticStateMachine when reinitialize()
https://github.com/apache/ratis/blob/dac27e4fd3df042fcf50ed549404e4d82708ed2b/ratis-examples/src/main/java/org/apache/ratis/examples/arithmetic/ArithmeticStateMachine.java#L82-L84
Will reset the lastAppliedIndex, I wonder if we can do this too? then the change will be simple

**Comment by leixm** [comment]
> https://github.com/apache/ratis/blob/dac27e4fd3df042fcf50ed549404e4d82708ed2b/ratis-examples/src/main/java/org/apache/ratis/examples/arithmetic/ArithmeticStateMachine.java#L82-L84

The latest snapshot file can not found in this UT.

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/StateMachine.java [hunk 1]
```diff
@@ -114,6 +98,7 @@ public void initialize(RaftServer server, RaftGroupId id, RaftStorage raftStorag
   public void reinitialize() throws IOException {
     LOG.info("Reinitializing state machine.");
     getLifeCycle().compareAndTransition(PAUSED, STARTING);
+    storage.refreshLatestSnapshot();
```
Comment (AngersZhuuuu) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/StateMachine.java:101]: Remove this

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/CelebornStateMachineStorage.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+  public static String getSnapshotFileName(long term, long endIndex) {
+    return SNAPSHOT_FILE_PREFIX + "." + term + "_" + endIndex;
+  }
+
+  @Override
+  public SingleFileSnapshotInfo getLatestSnapshot() {
+    final SingleFileSnapshotInfo s = latestSnapshot.get();
+    if (s != null) {
+      return s;
+    }
```
Comment (AngersZhuuuu) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/CelebornStateMachineStorage.java:192]: Seems we can directly remove this three line

**Comment by AngersZhuuuu** [comment]
Alluxio fix the same issue in https://github.com/Alluxio/alluxio/pull/12181

**Review by AngersZhuuuu — APPROVED** [review]
LGTM and pls check failed GA test

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/CelebornStateMachineStorage.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+import org.apache.ratis.statemachine.SnapshotRetentionPolicy;
+import org.apache.ratis.statemachine.StateMachineStorage;
+import org.apache.ratis.statemachine.impl.SingleFileSnapshotInfo;
+import org.apache.ratis.util.FileUtils;
+import org.apache.ratis.util.MD5FileUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Copy from org.apache.ratis.statemachine.impl.SimpleStateMachineStorage, We need to refresh latest
```
Comment (SteNicholas) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/CelebornStateMachineStorage.java:49]:
```suggestion
 * Copied from Apache Ratis {@link SimpleStateMachineStorage}, We need to refresh latest
```

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/CelebornStateMachineStorage.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+ * snapshot after install snapshot from leader which makes StateMachine load latest snapshot
+ * correctly.
+ */
+public class CelebornStateMachineStorage implements StateMachineStorage {
```
Comment (SteNicholas) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/CelebornStateMachineStorage.java:53]: Please add this file into `LICENSE`.

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/CelebornStateMachineStorage.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+import org.apache.ratis.statemachine.StateMachineStorage;
+import org.apache.ratis.statemachine.impl.SingleFileSnapshotInfo;
+import org.apache.ratis.util.FileUtils;
+import org.apache.ratis.util.MD5FileUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Copy from org.apache.ratis.statemachine.impl.SimpleStateMachineStorage, We need to refresh latest
+ * snapshot after install snapshot from leader which makes StateMachine load latest snapshot
```
Comment (SteNicholas) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/CelebornStateMachineStorage.java:50]:
```suggestion
 * snapshot after installing snapshot from leader which makes StateMachine load latest snapshot
```

**Review by RexXiong — APPROVED** [review]
LGTM, thanks!

**Review by SteNicholas — APPROVED** [review]
Overall LGTM. I left the minor comments for this pull request. BTW, could the copied class be removed after https://github.com/apache/ratis/pull/1111 released?

**Comment by leixm** [comment]
> Overall LGTM. I left the minor comments for this pull request. BTW, could the copied class be removed after [apache/ratis#1111](https://github.com/apache/ratis/pull/1111) released?

Of course it can be removed after [apache/ratis#1111](https://github.com/apache/ratis/pull/1111)  released.

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Comment by leixm** [comment]
Compile failed caused by [CELEBORN-1337](https://issues.apache.org/jira/browse/CELEBORN-1337)

**Comment by SteNicholas** [comment]
@leixm, please rebase the latest main branch.

**Comment by leixm** [comment]
All done.

**Review by SteNicholas — APPROVED** [review]

**Comment by SteNicholas** [comment]
Thanks. Merged to main(v0.6.0) and branch-0.5(0.5.1).

**Pull request closed without merging**